PSA: implement key derivation for ECC keys

mprse opened the pull request:
## Description

Resolves #3260

First proposition of the solution for derivation of ECC keys.

I created this PR to share my code with the team to confirm if this approach is correct.

The library builds, but I'm looking how to test these changes. It seems that there are no tests available for this scenario.

cc @gilles-peskine-arm 

## Status
**IN DEVELOPMENT**

## Requires Backporting

## Migrations
NO

## Additional comments

mprse commented:
@gilles-peskine-arm @mpg Can you please give some feedback for these changes?
- Does it make sense? 
- What is the best way to test key derivation for ECC keys?

mpg commented on library/psa_crypto.c:
```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
+    * 4. If k > N - 2, discard the result and return to step 1.
+    * 5. Output k + 1 as the private key.
+    *
+    * This method allows compliance to NIST standards
```
Can you give a precise reference? I'm not sure which standard I should check.

mpg replied:
Oh, OK, found the references in the PSA spec. Still, since we're quoting that spec extensively here, we might as well give the reference too (just the second bullet point, the one that's about ECC).

mprse replied:
This is taken from the `PSA Crypto API - psa_key_derivation_output_key (function) for ECC`:
https://armmbed.github.io/mbed-crypto/html/api/ops/kdf.html#psa_key_derivation_output_key

mpg commented on library/psa_crypto.c:
```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
+            /* 5. Output k + 1 as the private key. */
+            mbedtls_mpi sum_k_1;
+            mbedtls_mpi_init( &sum_k_1 );
+            mbedtls_mpi_add_int( &sum_k_1, &k, 1);
```
I think we can do this operation in place in order to save an MPI. If it were a normal variable, I'd agree, let's use a new name rather that mutate the variable, it's more readable and will generate the same code. But the more MPIs we have active at a given time, the more RAM we use, and again that's not something the compiler will fix for us, so we need to minimize the number of MPIs when we can.

mpg commented on library/psa_crypto.c:
```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
-    status = psa_key_derivation_output_bytes( operation, data, bytes );
-    if( status != PSA_SUCCESS )
-        goto exit;
+            mbedtls_ecp_keypair ecp;
```
Do we really need an `ecp_keypair` here, or can we get away with just an `mbedtls_ecp_group`?

mpg commented on library/psa_crypto.c:
```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
+            mbedtls_ecp_keypair ecp;
+            mbedtls_ecp_keypair_init( &ecp );
+
+            if( ( ret = mbedtls_ecp_group_load( &ecp.grp, grp_id ) ) != 0 )
```
Aw, it's pretty awkward that we have do to this in order get the value of `N` for this curve, but I don't see a better way. We should be doing this outside of the loop, though.

mpg commented on library/psa_crypto.c:
```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
+    *
+    * This method allows compliance to NIST standards
+    */
+    if ( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
```
I'm afraid we need to be more precise here: this procedure is not valid for Montgomery or Edwards curves, so we probably need to check the curve family here. Gilles can probably advise more easily than me.

mpg reviewed (changes requested):
The strategy looks good, but there are issues in the execution. Mostly, it's the fact that MPIs don't work like `int`s so code manipulating them needs to be written with different idioms.

Regarding testing, Gilles probably knows better than me if we, but even if we don't have test vectors there's still a few things we can do:
- at the very minimum, run the function and check that it returns success when it should
- check that it errors out cleanly if the `operation` passed has insufficient capacity
- generate keys with various curves, visually check the results, then use these results for known-answer tests - that's not as good as comparing to values from standards or other, known-good implementations, but at least it ensures that _if_ the function is correct now then it will remain correct in the future.

mprse commented:
Addressed most issues pointed by @mpg. Still open is problem with how to specify when the new algorithm applies (waiting for @gilles-peskine-arm input). 

Tested implementation using `key_ladder_demo` test program with the following modifications to test ECC keys:
```
przemek@przemek-VirtualBox:~/src/mbedtls/programs/psa$ git diff key_ladder_demo.c
diff --git a/programs/psa/key_ladder_demo.c b/programs/psa/key_ladder_demo.c
index cad875e01..698039fc8 100644
--- a/programs/psa/key_ladder_demo.c
+++ b/programs/psa/key_ladder_demo.c
@@ -124,7 +124,8 @@ int main( void )
 
 /* Size of the key derivation keys (applies both to the master key and
  * to intermediate keys). */
-#define KEY_SIZE_BYTES 40
+#define KEY_SIZE_BITES 521 // works for 384, 521 (for 225, 256 Key file too large)
+#define KEY_SIZE_BYTES ((KEY_SIZE_BITES + 7)/8)
 
 /* Algorithm for key derivation. */
 #define KDF_ALG PSA_ALG_HKDF( PSA_ALG_SHA_256 )
@@ -278,9 +279,9 @@ static psa_status_t derive_key_ladder( const char *ladder[],
 
     psa_set_key_usage_flags( &attributes,
                              PSA_KEY_USAGE_DERIVE | PSA_KEY_USAGE_EXPORT );
-    psa_set_key_algorithm( &attributes, KDF_ALG );
-    psa_set_key_type( &attributes, PSA_KEY_TYPE_DERIVE );
-    psa_set_key_bits( &attributes, PSA_BYTES_TO_BITS( KEY_SIZE_BYTES ) );
+    psa_set_key_algorithm( &attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256) );
+    psa_set_key_type( &attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1) );
+    psa_set_key_bits( &attributes, KEY_SIZE_BITES );
 
     /* For each label in turn, ... */
     for( i = 0; i < ladder_depth; i++ )
```

Command:
`./key_ladder_demo save label=hello input=../../tests/data_files/server6.key output=hello.key`

It successfully generates `hello.key` for key sizes 384, 521 bits, but I don't know if generated key is valid. These cases also covers positive and negative scenario for: `m is/is not a multiple of 8`.
For 225 and 256 bits I'm getting `Key file too large` error while importing key from file.

@mpg Ready for the next review round.

mprse commented:
Got a problem with finding solution for the following CI failure:
```
./tests/scripts/test_psa_compliance.py

The following test case is failing:

******************************************

TEST: 221 | DESCRIPTION: Testing crypto key derivation APIs | UT: psa_key_derivation_output_key
[Info] Executing tests from non-secure
[Check 1] Test psa_key_derivation_output_key - Key
[Check 2] Test psa_key_derivation_output_key - Info
[Check 3] Test psa_key_derivation_output_key - Salt
[Check 4] Test psa_key_derivation_output_key - Greater Capacity than available
[Check 5] Test psa_key_derivation_output_key - ECC Public key
[Check 6] Test psa_key_derivation_output_key -  ECC keypair
	Failed at Checkpoint: 14
	Actual: 0
	Expected: -135

TEST RESULT: FAILED (Error Code=0x1)
ERROR: Unexpected failure

******************************************
```

This test case is related to my changes, but I don't know where this test is located. It probably needs to be adapted.

cc @mpg @gilles-peskine-arm

mprse commented:
Ok, I see now. 
This test is failing: https://github.com/bensze01/psa-arch-tests/blob/master/api-tests/dev_apis/crypto/test_c021/test_c021.c

mprse commented:
```
[Check 6] Test psa_key_derivation_output_key -  ECC keypair
	Failed at Checkpoint: 14
	Actual: 0
	Expected: -135
```

Check 14 fails because test with these changes key derivation for ECC keypair succeeds. 

After changing expected result I end up with:
```
[Check 6] Test psa_key_derivation_output_key -  ECC keypair
	Failed at Checkpoint: 17
	Actual: 0
	Expected: -143
```
I assume that second derivation should fail, so there is some problem with implementation. Unfortunately have no idea what is missing.

mpg commented:
> When I used `if ( slot->attr.policy.alg == PSA_ALG_ECDSA_ANY )` instead `PSA_KEY_TYPE_IS_ECC( slot->attr.type )` the `Test psa_key_derivation_output_key - RSA keypair` test has passed. But seems that this is not optimal solution.

Indeed that sound very weird. I'm not very familiar with that part of the code but looking at the existing code it looks to me like this shouldn't make a difference to the result of the RSA test. (None of the conditions is exactly the right one, as we want to target only Short Weierstrass curves, but that's a different issue and shouldn't affect any RSA test.) @gilles-peskine-arm do you think you could have a look?

gilles-peskine-arm commented:
> Test psa_key_derivation_output_key - RSA Public Key

That's a public key, so attempting to derive it should keep failing with `INVALID_ARGUMENT`.

> Test psa_key_derivation_output_key - RSA keypair

That one should keep failing, although `NOT_SUPPORTED` would be a better error code (it's technically possible but we don't implement it). The problem here is that there's a mistake in the test data.

https://github.com/ARM-software/psa-arch-tests/pull/294

@bensze01 I made a PR to fix a bug in the PSA arch tests. Should I submit this PR to your fork as well?

mprse commented:
Confirmed that the following change fixes my problem with failing `Test psa_key_derivation_output_key -  RSA keypair` test case:
```
- PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1), 2046, PSA_KEY_USAGE_EXPORT,
+ PSA_KEY_TYPE_RSA_KEY_PAIR, 2048, PSA_KEY_USAGE_EXPORT,
```

I'm not sure if changing `INVALID_ARGUMENT -> NOT_SUPPORTED` is an option if we don't change the implementation (both test cases now are passing with `INVALID_ARGUMENT`).

Below updated version of `Test psa_key_derivation_output_key -  ECC keypair` adapted for this PR:
https://github.com/bensze01/psa-arch-tests/pull/1

bensze01 commented:
Ok, my plan for keeping the mbedtls branches / PRs and our fork of the psa-arch-tests in sync is the following:
1. create new branches `mbedtls-3` and `mbedtls-2` (same commits as `fixes-for-mbedlts-{3,2}`)
2. convert the `fixes-for-mbedtls-{2,3}` branches into tags
3. merge @mprse's PR into the `mbedtls-3` branch
4. tag the merge commit as `fix-pr-5139`
5. change the compliance script in this PR to checkout `fix-pr-5139` from the fork.

bensze01 commented:
On second thought, creating just a single branch (`mbedtls`) is probably best, since then we don't need to create separate tags for the development_2.x fixes, if any come up in the future.

I'll also probably just leave the existing branches be as static branches instead of converting them to tags.

mprse commented:
@bensze01 At the moment we are checking out always the head of the 'fixes-for-mbedtls-3' branch (not by tag):
https://github.com/ARMmbed/mbedtls/blob/development/tests/scripts/test_psa_compliance.py#L71

It would be good to checkout by tag in the script because then we can change the tag in the specific PR, so it still works fine for other PRs.

bensze01 commented:
@mprse Yes, that's what I'm planning to do. I still intend to only check out tags going forward, starting with this PR. I was just going to skip the (optional) 2nd step in my original comment, and just not touch those branches from now on.

I don't think many people have cloned my fork (aside from the script included in mbedtls, but the script should be resilient against this), but git could complain about tag-branch collisions on their end if they did, and I replace the branches with tags.

Anyways, I comfortable with either solution.

bensze01 commented:
@gilles-peskine-arm I think I'll cherry pick your first and third commit sfrom that PR.  
The middle commit causes a new failure with our current code, that is not in scope for this PR to fix.

bensze01 commented:
Actually, I'll just go ahead and merge the first commit. The third commit is a nice to have, but not necessary for us to pass the test.

bensze01 commented:
@mprse I've merged your PR on my fork and tagged the merge commit.

Please rebase this PR on top of development, and edit `test_psa_compliance.py` to checkout `fix-pr-5139` instead of `fixes-for-mbedtls-3`.

mprse commented:
Rebased on top of `development` and adapted `test_psa_compliance.py` (https://github.com/ARMmbed/mbedtls/pull/5139/commits/6d3cdc1cdf61096503cf4829d8984514d7e7d770).

Tested locally:
```
************ Crypto Suite Report **********
TOTAL TESTS     : 63
TOTAL PASSED    : 52
TOTAL SIM ERROR : 0
TOTAL FAILED    : 11
TOTAL SKIPPED   : 0
******************************************

Entering standby.. 

***** test_psa_compliance.py report ******

Expected failures: 252, 253, 254, 255, 256, 257, 258, 259, 261, 262, 263
Unexpected failures: 
Unexpected successes: 

SUCCESS
```

Hope that now CI will pass 🤞 
@mpg @gilles-peskine-arm @bensze01 Thanks for help!

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4695,129 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
 #endif /* MBEDTLS_PSA_BUILTIN_KEY_TYPE_DES */
 
+/*
+* ECC key types require the generation of a private key which is an integer
```
Note that this is just the method for Weierstrass curves. Montgomery curves use a different method.

mprse replied:
Fixed

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4695,129 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
 }
 #endif /* MBEDTLS_PSA_BUILTIN_KEY_TYPE_DES */
 
```
In https://github.com/ARMmbed/mbedtls/pull/5218 I wrote a script that can generate expected output for test cases.

Since you haven't pushed a new test function yet, I wrote my own. If you have a different test function in your working copy I or you can adapt my code.

mprse replied:
I was testing the implementation as described [here](https://github.com/ARMmbed/mbedtls/pull/5139#issuecomment-972778279).

Thanks I'll try to analyse and use your script.

gilles-peskine-arm replied:
I've added ECC test cases, including a Weierstrass test case that requires redrawing.

gilles-peskine-arm commented:
You labeled the PR "needs: review", but it's incomplete (no tests). What's the status now? Do you need a design review?

mprse commented:
Well https://github.com/ARMmbed/mbedtls/issues/3260 does not say about tests. 

I know I need to test somehow the implementation, but I didn't know how to verify if generated key is correct. I only tested using `key_ladder_demo` and debugger that code is executed as expected.

gilles-peskine-arm commented:
Unless otherwise specified, any feature task includes coding the feature itself, testing it and documenting it. Any bug-fix task includes fixing the bug and adding a non-regression test.

Here the documentation is just a changelog entry. For testing, I think https://github.com/ARMmbed/mbedtls/pull/5218 now has enough known-answer test cases. We'll need a few postiive test cases to cover _using_ the derived key (i.e. set an appropriate policy and call `mbedtls_test_psa_exercise_key` to verify that the key can be used for signing or key agreement — `derive_key_exercise` does that), and a few more negative test cases to have sufficient coverage of key types.

mprse commented:
So what is the next step in this case? Should I bring test code + test vectors from #5218 to this PR?

gilles-peskine-arm commented:
Yes, you might as well take them, and adapt them if needed.

mprse commented:
Moved test function and test vectors from  #5218 and enabled tests for ECC (not Montgomery). Confirmed that code is covered (`m % 8` and re-draw) by tests. PR is completed and ready for review.

I'm not sure about operation capacity and re-draw operation. Capacity will be decreased twice, is that ok?

gilles-peskine-arm commented:
> I'm not sure about operation capacity and re-draw operation. Capacity will be decreased twice, is that ok?
Yes, that's expected.

mprse commented:
Added support for Montgomery Curves and enabled tests. 

Montgomery curves support only `255, 448` key bits. I got check for this, but check for Weierstrass curves is missing. 

Do I need to verify if `bits` are correct for specific curve (`PSA_ECC_FAMILY_SECP_K1, PSA_ECC_FAMILY_SECP_R1, PSA_ECC_FAMILY_SECP_R2, PSA_ECC_FAMILY_SECT_K1, PSA_ECC_FAMILY_SECT_R1, PSA_ECC_FAMILY_SECT_R2, PSA_ECC_FAMILY_BRAINPOOL_P_R1, PSA_ECC_FAMILY_FRP`) like described [here](https://armmbed.github.io/mbed-crypto/html/api/keys/types.html#c.psa_ecc_family_t)? 
Do we support all of those curves (in generated test vectors only `PSA_ECC_FAMILY_SECP_R1` curve is tested and also not all possible `bits` for this curve) ?

gilles-peskine-arm commented:
For Weierstrass curves, the calculation depends on the order of the curve. If the requested combination of curve family and size isn't supported, the library can't know the order of the curve, so this leads to an error case, which must be tested.

mprse commented on library/psa_crypto.c:
```diff
@@ -4820,6 +4820,50 @@ static psa_status_t psa_generate_derived_ecc_key_weierstrass_helper(
+                return ( PSA_ERROR_INVALID_ARGUMENT );
+            break;
+/*
+        case ( PSA_ECC_FAMILY_FRP ):
```
`PSA_ECC_FAMILY_FRP` is defined in the PSA documentation but not in code.

mprse commented:
- added implementation for Montgomery curves
- enabled test vectors for Montgomery curves
- added curve/bits consistency check + negative test vectors (not sure if this is enough)
- added test vectors for ECC key exercise when deriving an ECC key
- fixed issues pointed by CI

Setting to ready for review.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -4693,6 +4693,22 @@ PSA key derivation: TLS 1.2 PRF SHA-256, exercise HKDF-SHA-256
 depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_ALG_TLS12_PRF
 derive_key_exercise:PSA_ALG_TLS12_PRF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8f9":PSA_KEY_TYPE_DERIVE:400:PSA_KEY_USAGE_DERIVE:PSA_ALG_HKDF(PSA_ALG_SHA_256)
 
+PSA key derivation: HKDF-SHA-256 -> ECC secp256r1, exercise PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1)-CTR
```
As below, please make the first 66 characters of the test descriptions unique.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -4693,6 +4693,22 @@ PSA key derivation: TLS 1.2 PRF SHA-256, exercise HKDF-SHA-256
 
+PSA key derivation: HKDF-SHA-256 -> ECC secp256r1, exercise PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1)-CTR
+depends_on:PSA_WANT_ALG_CTR:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256
+derive_key_exercise:PSA_ALG_HKDF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8f9":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1):256:PSA_KEY_USAGE_DERIVE:PSA_ALG_CTR
```
There has to be something wrong here: `PSA_ALG_CTR` is not an algorithm that you can use on an ECC key. What's happening is that `exercise_key` only does what is permitted by the triplet (key type, algorithm, usage flags) and doesn't complain if that triplet is fishy. (I'm not sure `exercise_key` should complain here: it's not really its job to validate test data, and there may well be cases where this behavior is desired, to test keys with attributes that are technically correct but not really useful.)

The following policy would make sense (adding `EXPORT` so as to test the export of the key as well):
```suggestion
derive_key_exercise:PSA_ALG_HKDF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8f9":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1):256:PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY:PSA_ALG_ECDSA
```
and update the description and dependencies accordingly.

For Curve25519 keys, use `PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_DERIVE:PSA_ALG_ECDH`

mprse replied:
Do you mean the following change:
`PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_SIGN_HASH | PSA_KEY_USAGE_VERIFY_HASH:PSA_ALG_ECDSA_ANY`

gilles-peskine-arm replied:
Yes, right (for the Weierstrass test cases).

mprse commented on library/psa_crypto.c:
```diff
@@ -4840,14 +4884,20 @@ static psa_status_t psa_generate_derived_key_internal(
         {
             /* Weierstrass elliptic curve */
             unsigned key_err = 0;
+            status = psa_generate_derived_ecc_key_weierstrass_check_config(
```
Yes, this function seems redundant as `mbedtls_ecc_group_of_psa` can be used instead and the check can be done in `psa_generate_derived_ecc_key_weierstrass_helper`.

But the proposed condition for invalid argument seems to be invalid.
```
if( mbedtls_ecc_group_of_psa( curve, bits, 0 ) == MBEDTLS_ECP_DP_NONE ) 
    return( PSA_ERROR_INVALID_ARGUMENT )
```

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,179 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+    size_t m = ecp_group.nbits;
+
+    size_t m_bytes = PSA_BITS_TO_BYTES( m );
+    if (*data == NULL)
+        *data = mbedtls_calloc( 1, m_bytes );
```
This function determinates the required size of the buffer and allocates the buffer. This buffer is then used by the caller and released by the caller in positive scenario. 
In case of error path buffer is released by this function and pointer set to `NULL`. Caller will in this case free null pointer which is ok according to C standard (`free(NULL)` has no effect). 
The conditional allocation is done because this function can be called second time when value is out of range. In this case we want to use buffer that is already allocated and just fill it with the new data.

gilles-peskine-arm replied:
Now that `psa_generate_derived_ecc_key_weierstrass_helper` is not called in a loop, `*data` is always `NULL` at this point, right? So now it would make sense for `psa_generate_derived_ecc_key_weierstrass_helper` to allocate unconditionally, because that's the place where you know the size, and it makes sense to use a heap buffer because that buffer will end up in the key store.

This part looks unfinished, that's why I thought you were still working on this PR and I didn't re-review it. Do you think there's still a reason for the `if (*data == NULL)` check? At a cursory glance (I have NOT actually reviewed yet), you could just remove this line and document that the function sets `*data` to memory that it allocates.

Actually, please document the behavior of the function regarding `*data` in any case.

mprse replied:
Yes, `psa_generate_derived_ecc_key_weierstrass_helper` is now always called with `*data == NULL`, so `(*data == NULL)` is redundant. I will remove it and add comment.

I still didn't address previous comment with request to move `data` buffer to stack. It didn't work for me and I end up with the following failure:
`PSA key derivation: largest possible key .......................... *** stack smashing detected ***: terminated`.
I suspect that local data buffer of size `PSA_BITS_TO_BYTES( PSA_VENDOR_ECC_MAX_CURVE_BITS )` is ok for ECC case but it is not enough for other key types.

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,179 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+         * (8 * ceiling(m/8) - m) bits of the first byte in
+         * the string to zero.
+         */
+        uint8_t clear_bit_count = (uint8_t) ( 8 * m_bytes - m );
+        uint8_t clear_bit_mask = ( ( 1 << clear_bit_count ) - 1 );
+        clear_bit_mask = ~( clear_bit_mask << ( 8 - clear_bit_count ) );
+        *data[0] = ( *data[0] & clear_bit_mask );
```
This will clear the least (instead of most) significant (8 * ceiling(m/8) - m) bits of the first byte in the string to zero.

mprse replied:
This will clear the (8 - bits to clear) most significant bits.

gilles-peskine-arm replied:
How so? It looks right to me. Larger m means more bits are kept.

mprse commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -4717,15 +4825,95 @@ PSA key derivation: invalid type (PSA_KEY_TYPE_CATEGORY_MASK)
 depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256
 derive_key:PSA_ALG_HKDF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8f9":PSA_KEY_TYPE_RAW_DATA:7:PSA_ERROR_INVALID_ARGUMENT:0
 
+PSA key derivation: invalid length PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1) (0)
```
> Maybe we should make it an error if the 66-character prefix is non-unique?

I agree that this would be helpful.

gilles-peskine-arm replied:
Filed as https://github.com/ARMmbed/mbedtls/issues/5257

mprse commented:
I got here strange error on CI during build on Windows:

```
    65>c:\builds\workspace\mbed-tls-pr-head_pr-5139-head\worktrees\tmprcpaquas\library\psa_crypto.c(4838): fatal error C1001: An internal error has occurred in the compiler. [C:\builds\workspace\mbed-tls-pr-head_PR-5139-head\worktrees\tmprcpaquas\cmake_solution\library\mbedcrypto.vcxproj]
         (compiler file 'f:\dd\vctools\compiler\utc\src\p2\main.c', line 228)
          To work around this problem, try simplifying or changing the program near the locations listed above.
         Please choose the Technical Support command on the Visual C++ 
          Help menu, or open the Technical Support help file for more information
         Internal Compiler Error in C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\CL.exe.  You will be prompted to send an error report to Microsoft later.
```

CI passed before I addressed review changes. Have someone seen such error before? Or can give some hint how to proceed with this?
cc @bensze01 @gilles-peskine-arm

mprse commented:
```
+ ./tests/scripts/all.sh --seed 4 --keep-going test_psa_crypto_config_accel_cipher


psa_crypto.c: In function 'libtestdriver1_psa_generate_derived_ecc_key_weierstrass_helper':

psa_crypto.c:4790:9: error: implicit declaration of function 'libtestdriver1_mbedtls_ecc_group_of_psa' [-Werror=implicit-function-declaration]

         libtestdriver1_mbedtls_ecc_group_of_psa( curve, bits, 0 );
```

I need to rebase this one as above build fails and I don't have `test_psa_crypto_config_accel_cipher` in sources.

mprse commented:
I solve the new build issue without re-basing, but unfortunately error during windows build still exists.

bensze01 reviewed (changes requested):
@mprse A heads up: #5388 changed the compliance suite tag we test against, so to resolve the conflict in `test_psa_compliance.py` you will need to do a merge / rebase and change the test suite tag from `fix-pr-5139` to `fix-pr-5139-1` .

mprse commented:
Thanks @bensze01 for the details. 

@mpg @gilles-peskine-arm Should I rebase this one now or wait for re-review?

gilles-peskine-arm commented:
I have a review in progress (must finish it...), can you please wait?

mprse commented:
> I have a review in progress (must finish it...), can you please wait?

Sure 👍

bensze01 commented:
In the meantime, the upstream compliance suite tag changed again. `fix-pr-5139-2` is now the correct tag for this PR.

I don't have any more compliance suite PRs in flight, so I don't expect any more merge conflicts on that front.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4881,26 +4832,21 @@ static psa_status_t psa_generate_derived_key_internal(
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDH)
     if ( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
     {
         if ( PSA_KEY_TYPE_ECC_GET_FAMILY( slot->attr.type ) != PSA_ECC_FAMILY_MONTGOMERY )
```
Please avoid `if (condition expressed negatively) ... else ...`. Generally, please avoid double negations: it gets hard to keep track of them (was it not equal, or not not equal, or not not not equal?).

One of these days we'll add Edwards curves in addition to Montgomery and Weierstrass. So I think we'll need to introduce a macro `PSA_ECC_FAMILY_IS_WEIERSTRASS` (and maybe `IS_MONTGOMERY` as well, but this one is likely to remain forever implemented as just `== PSA_ECC_FAMILY_MONTGOMERY`). Now would be a good time to introduce this macro.

The implementation of the macro can use the structure of family encodings. For the foreseeable future, a correct implementation would be `(family & 0xc0) == 0`.

mprse replied:
I can add macro `PSA_ECC_FAMILY_IS_WEIERSTRASS`.

I see only the following curve families in code:

```
#define PSA_ECC_FAMILY_SECP_K1           ((psa_ecc_family_t) 0x17)
#define PSA_ECC_FAMILY_SECP_R1           ((psa_ecc_family_t) 0x12)
#define PSA_ECC_FAMILY_SECP_R2           ((psa_ecc_family_t) 0x1b)
#define PSA_ECC_FAMILY_SECT_K1           ((psa_ecc_family_t) 0x27)
#define PSA_ECC_FAMILY_SECT_R1           ((psa_ecc_family_t) 0x22)
#define PSA_ECC_FAMILY_SECT_R2           ((psa_ecc_family_t) 0x2b)
#define PSA_ECC_FAMILY_BRAINPOOL_P_R1    ((psa_ecc_family_t) 0x30)
#define PSA_ECC_FAMILY_MONTGOMERY        ((psa_ecc_family_t) 0x41)
#define PSA_ECC_FAMILY_TWISTED_EDWARDS   ((psa_ecc_family_t) 0x42)
```

There is no WEIERSTRASS family defined. So how do we know that the following condition can be used to determine if curve is Weierstrass `(family & 0xc0) == 0` ?

mpg replied:
All of these families are Weierstrass curves except the last two (Montgomery and Edwards). Historical note: initially all standardized curves were Weierstrass, so that's not reflected in the family names, because it was assumed. Montgomery and Edwards are more recent additions to the world of ECC.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4705,21 +4879,90 @@ static psa_status_t psa_generate_derived_key_internal(
+                    data[55] |= 128;
+                    break;
+                default:
+                    /* already handled */
```
Please update this comment: it's misleading.

mprse commented:
> Given that this buffer is allocated within psa_generate_derived_key_internal and freed when the function returns, and given that its size has a known, maximum bound, please make it a stack buffer. The size bound is PSA_BITS_TO_BYTES( PSA_VENDOR_ECC_MAX_CURVE_BITS ) (or if you prefer you can express it as PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE( PSA_VENDOR_ECC_MAX_CURVE_BITS )).

This buffer is used also in case of non ECC key. Is `PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE( PSA_VENDOR_ECC_MAX_CURVE_BITS ))` length also ok for this case?

mprse commented:
Addressed all review comments apart from moving `data` buffer to stack.

mprse commented:
Can we move forward with this one? @gilles-peskine-arm I addressed all the review comments (except from moving data buffer to stack). When the PR is accepted I can rebase it and push through CI. Should be ready then.

mpg commented:
I can make time to review this PR this week - it's been waiting for too long, sorry.

mpg commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+        *    Result of comparison is returned. When it indicates error
+        *    then this fuction is called again.
+        */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_int( &diff_N_2, &ecp_group.N, 2) );
```
This can be computed outside of the loop.

mprse replied:
Moved outside the loop.

mpg commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+        *    then this fuction is called again.
+        */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_int( &diff_N_2, &ecp_group.N, 2) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_grow( &k, diff_N_2.n ) );
```
Note: with the current implementation of `mbedtls_mpi_read_binary()`, this is not necessary. However, that's currently not documented. I think it would be worth adding to the internal documentation of `mbedtls_mpi_read_binary()` (that is, in `bignum.c`, but the comment above the function, to make it clear it's part of the contract) what the number of limbs of the result is, so we can start relying on it.

mprse replied:
Should I remove `MBEDTLS_MPI_CHK( mbedtls_mpi_grow( &k, diff_N_2.n ) );` ?

mpg replied:
I think so, but before doing that we need to make it part of the (internal) contract of `mbedtls_mpi_read_binary()` by changing the comment in `bignum.c` to something like:
```
 * Import X from unsigned binary data, little endian
 *
 * This function is guaranteed to return an MPI with exactly the necessary
 * number of limbs (in particular, it does not skip 0s in the input).
```

gilles-peskine-arm replied:
Oh, yes we should! IIRC there's library code that depends on that for constant-timeness, and there's definitely test code that depends on that for testing what it's supposed to test (operations on `mbedtls_mpi` objects with leading zero limbs).

gilles-peskine-arm replied:
Well, more precisely, there's code that depends on `read_binary` keeping leading zeroes in the result. I'm not sure there's anything that depends on it shrinking the output object if it's already larger than it needs to be.

mprse replied:
Seem that the same comment can be also added to `mbedtls_mpi_read_binary_le()`?

mpg replied:
> I'm not sure there's anything that depends on it shrinking the output object if it's already larger than it needs to be.

This PR does. (Even before the last commit, I mean.) `mbedtls_mpi_lt_mpi_ct()` only works when its two arguments have the same allocated size. (So, actually, if we didn't want to rely on `read_binary()` doing what we want, we should have called `mbedtls_mpi_resize_clear()` not `mbedtls_mpi_grow()`.

mpg commented on library/psa_crypto.c:
```diff
@@ -4705,21 +4833,81 @@ static psa_status_t psa_generate_derived_key_internal(
+        else
+        {
+            /* Montgomery elliptic curve */
+            size_t output_length;
```
Minor: I think it would be better to make that a helper function too. It's OK as it is now, but as we add support for more key types we don't want this function to end up being hundreds of lines, so it's probably better if it just dispatches to specialized helpers.

mprse replied:
Added helper function for Montgomery curves.

mpg commented on library/psa_crypto.c:
```diff
@@ -4705,21 +4833,81 @@ static psa_status_t psa_generate_derived_key_internal(
+#endif
+    {
+        if( ! key_type_is_raw_bytes( slot->attr.type ) )
+            return( PSA_ERROR_INVALID_ARGUMENT );
```
Minor: currently the structure is
```
if( is_ecc ) {
   /* handle ecc */
} else {
    if( !is_raw_bytes )
       return( invalid );
    /* handle raw bytes */
}
```
I think I'd find the following clearer:
```
if( is_ecc ) {
   /* handle ecc */
} else if( !is_raw_bytes ) {
    /* handle raw bytes */
} else
    return( invalid );
}
```
(more readable when more type are added)

mprse replied:
Refactored if-else statement as suggested.

mpg commented on library/psa_crypto.c:
```diff
@@ -4705,21 +4833,81 @@ static psa_status_t psa_generate_derived_key_internal(
+        status = psa_key_derivation_output_bytes( operation, data, bytes );
+        if( status != PSA_SUCCESS )
+            goto exit;
 #if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_DES)
```
Pre-existing: this condition only has `BUILTIN` while the ECC case has `BUILTIN || ACCEL`. Is the existing condition correct? @gilles-peskine-arm

mpg commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -4709,6 +4717,98 @@ PSA key derivation: TLS 1.2 PRF SHA-256, derive key export, 1+41
+depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_521
+derive_key_type:PSA_ALG_HKDF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8fa":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1):521:"01122f37d10965c8455ecbd2bc73d5da5347d0ce772e54305d528295a64ffb7c567f5042e2d7e5803b407c08d1e110adcefc35564035d706582f723a2f76a32260da"
+
+# For Curve25519, test a few different outputs to exercise masking.
```
Can you expand the comment to mention that the variation is in the last byte of `input_2`? Took me some time to find it ^^ (Same for Curve448 below.)

mprse replied:
Added comment.

mpg commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -4709,6 +4717,98 @@ PSA key derivation: TLS 1.2 PRF SHA-256, derive key export, 1+41
 
+PSA key derivation: HKDF-SHA-256 -> AES-128
+depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256:PSA_WANT_KEY_TYPE_AES
+derive_key_type:PSA_ALG_HKDF(PSA_ALG_SHA_256):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":"000102030405060708090a0b0c":"f0f1f2f3f4f5f6f7f8f9":PSA_KEY_TYPE_AES:128:"3cb25f25faacd57a90434f64d0362f2a"
```
Question: where do the `expected_export` values come from? I assume they were just generated with Mbed TLS?

mprse replied:
The expected results were generated using the python script provided by @gilles-peskine-arm in PR https://github.com/ARMmbed/mbedtls/pull/5218.

mprse replied:
It is mentioned in the commit message https://github.com/ARMmbed/mbedtls/pull/5139/commits/cb3eacd6913a721f348f73247fffbfa77ce2dd01.

mpg replied:
Oh sorry, I missed it, my fault.

mpg commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6375,6 +6375,65 @@ exit:
 /* END_CASE */
 
+/* BEGIN_CASE */
+void derive_key_type( int alg_arg,
```
I'm not sure why this function's name ends with `_type`. It seems to me that what differentiates it from `derive_key` is the presence of the `expected_export` argument, to it seems to me that `derive_key_result` would probably be a better name.

mprse replied:
To be honest I don't remember why test function name ends with `_type`. Will change as suggested.

mprse replied:
While looking for the comment about generating test vectors I noticed that function code was imported also from https://github.com/ARMmbed/mbedtls/pull/5218.

mpg replied:
Ah OK, so let's keep it unchanged here, and discuss possible changes in #5218 instead then.

gilles-peskine-arm replied:
This function is about deriving keys of different types, so the name makes sense to me.

mprse replied:
According to the comment in https://github.com/ARMmbed/mbedtls/pull/5139/commits/cb3eacd6913a721f348f73247fffbfa77ce2dd01 test function was taken from https://github.com/ARMmbed/mbedtls/pull/5218.

mpg commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6375,6 +6375,65 @@ exit:
+            input1->x, input1->len,
+            input2->x, input2->len,
+            PSA_KEY_DERIVATION_UNLIMITED_CAPACITY ) )
+        goto exit;
```
Question: why not use `PSA_ASSERT()` here? If there's a reason, it's not obvious to me, so I think a comment would help.

mprse replied:
Seems that `PSA_ASSERT(...)` is not suitable for this case as it compares result with `PSA_SUCCESS` and `mbedtls_test_psa_setup_key_derivation_wrap()` returns `int` not `psa_status_t`. Maybe other `ASSERT` could be used, but all other test functions use the same condition, so for consistency I suggest to leave it as it is now.

mpg replied:
Ok, thanks, I didn't notice it was returning `int`. Agreed.

mpg reviewed (changes requested):
Looks pretty good to me, I think the only issues I found are either style, or requests for clarification (comments / naming / code structure), but other than that the code and tests are looking good.

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+    mbedtls_ecp_group ecp_group;
+    mbedtls_ecp_group_init( &ecp_group );
+
+    if( ( status = mbedtls_to_psa_error( mbedtls_ecp_group_load( &ecp_group, grp_id ) ) ) != 0 )
```
Fixed

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+    mbedtls_mpi k;
+    mbedtls_mpi diff_N_2;
+    /* ret variable is initialized to 0 as it is
+       used only by MBEDTLS_MPI_CHK macro */
```
Initialized `ret` to error value adapted code.

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+    size_t m_bytes = PSA_BITS_TO_BYTES( m );
+
+    /* Note: This function is always called with *data == NULL and it
+     * allocates memory for the data buffer. */
```
Added note to the function description.

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+            goto cleanup;
+
+        /* 2. If m is not a multiple of 8 */
+        if (m % 8)
```
Fixed code style.

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+        /* 3. Convert the string to integer k by decoding it as a
+        *    big-endian byte string.
+        */
+        MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary( &k, *data, m_bytes));
```
Fixed

mprse commented on library/psa_crypto.c:
```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+        if (m % 8)
+        {
+            /* Set the most significant
+            * (8 * ceiling(m/8) - m) bits of the first byte in
```
Fixed

mprse commented:
@mpg Review comments addressed. CI passes.

mpg reviewed (approved):
Thanks for addressing my feedback. I'm happy with the result. Unfortunately a conflict appeared. Please check with @gilles-peskine-arm when's good time for you to rebase without disrupting his re-review.

gilles-peskine-arm commented:
@mprse Please go ahead and rebase.

mprse commented:
> @mprse Please go ahead and rebase.

Ok. What should I do with the test function and test vectors taken from https://github.com/ARMmbed/mbedtls/pull/5218? Should this be part of this PR and 5218 will provide the gen script only?

mprse commented:
Rebased on top of development with modified tag in `test_psa_compliance.py` to `fix-pr-5139-2`.

gilles-peskine-arm commented:
> Ok. What should I do with the test function and test vectors taken from #5218? Should this be part of this PR and 5218 will provide the gen script only?

Please keep the tests (code and vectors) here. The generation script in 5218 can wait, and it doesn't matter if the test vectors we actually use diverge from what the script generates.

mprse commented:
Rebased and CI has passed.

mpg reviewed (dismissed):
Thanks for rebasing. I'm happy with the result (checked the new commit as well).

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4724,20 +4724,22 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
     defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY) || \
-    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR) || \
-    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_PUBLIC_KEY)
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || \
```
Preexisting: the conditions on `mbedtls_ecc_group_of_psa` need to be changed as well. Since it's a public API function now, it should be defined whenever it makes sense, which is `defined(MBEDTLS_ECP_C) && defined(PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY)`. (Or `defined(PSA_WANT_ECC_KEY_TYPE_ECC_KEY_PAIR)`, but this implies that `xxx_PUBLIC_KEY` is also defined, so it doesn't need to be checked.)

Still an issue here: whether ECDSA and ECDH are supported is not relevant for ECC key derivation. The only condition here should be `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR`. (Not `PUBLIC_KEY` since you can't derive a public key.)

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4824,6 +4825,200 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+    mbedtls_mpi k;
+    mbedtls_mpi diff_N_2;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
```
```suggestion
    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
```

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4834,21 +5029,48 @@ static psa_status_t psa_generate_derived_key_internal(
-    data = mbedtls_calloc( 1, bytes );
-    if( data == NULL )
-        return( PSA_ERROR_INSUFFICIENT_MEMORY );
+#if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR) || \
+    defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDH)
```
As above:
```suggestion
#if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR)
```

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4824,6 +4825,200 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+            (*data)[0] &= 252;
+            (*data)[55] |= 128;
+            break;
+        default:
+            /* should never happen */
+            break;
```
It doesn't make much sense to do explicitly nothing in a can't-happen case. Please either wipe the output buffer and return `PSA_ERROR_CORRUPTION_DETECTED` the `default` case.

gilles-peskine-arm reviewed (dismissed):
Thanks for the updates. Looks good to me!

mprse commented:
Changed build flags (https://github.com/ARMmbed/mbedtls/pull/5139/commits/f110dc05beb1f6df67e3f86adaaa5686e4f38d91) are now causing build errors on CI in few configurations.

`mbedtls_psa_ecp_load_representation()` is built if any of the following flags is defined:
- `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR`
- `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY`
- `MBEDTLS_PSA_BUILTIN_ALG_ECDSA`
- `MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA`
- `MBEDTLS_PSA_BUILTIN_ALG_ECDH`

`mbedtls_psa_ecp_load_representation()` calls `mbedtls_ecc_group_of_psa()` which is now built only under `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR` flag.

Tried to change `mbedtls_psa_ecp_load_representation()` to be built also under `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR` only, but this implies build errors in functions that call `mbedtls_psa_ecp_load_representation()`. 

cc @gilles-peskine-arm

mprse commented on tests/scripts/test_psa_compliance.py:
```diff
@@ -47,7 +47,7 @@
 PSA_ARCH_TESTS_REPO = 'https://github.com/bensze01/psa-arch-tests.git'
-PSA_ARCH_TESTS_REF = 'fix-pr-5272'
+PSA_ARCH_TESTS_REF = 'fix-pr-5139-2'
 
```
Yes this is the second issue. Seems that there are two options: adapt compliance tests or revert the change. I we want to go with `PSA_ERROR_NOT_SUPPORTED` I think I  need to create PR to https://github.com/bensze01/psa-arch-tests.git. Is that correct @bensze01 ?

bensze01 replied:
I've rebased the fixes on top of the upstream master, and tagged the result of the merge. I'll re-trigger the CI now.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -324,11 +325,7 @@ psa_status_t mbedtls_to_psa_error( int ret )
-    defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) || \
-    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDH)
+#if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR)
 mbedtls_ecp_group_id mbedtls_ecc_group_of_psa( psa_ecc_family_t curve,
```
(continuing https://github.com/ARMmbed/mbedtls/pull/5139#discussion_r816289534, https://github.com/ARMmbed/mbedtls/pull/5139#discussion_r777546822, https://github.com/ARMmbed/mbedtls/pull/5139#issuecomment-1057821528, but you don't need to read those to follow this discussion)

If neither `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR` nor `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY` is defined, this indicates that the built-in crypto does not support any elliptic curve key, and so `mbedtls_ecc_group_of_psa` can only ever return 0.  A configuration in which, for example, neither `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR` nor `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY` are defined, but `MBEDTLS_PSA_BUILTIN_ALG_ECDSA` is defined, is not useful: ECDSA can only be used with ECC keys, and you can't ever perform ECDSA if you can't create an ECC key.

Hence `defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR) || defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY)` is the right condition here. (Earlier I stated that this simplifies to `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY`, since you can't have private key support without public key support; but this is true for `WANT` symbols, but not necessarily for `BUILTIN` symbols since it's in principle possible to have an accelerator for ECC public key operations only, and thus the builtin ECC code would only be used for private operations.)

However, there are preexisting issues with the test scripts for which I do not have a simple solution. So I propose to keep the original condition for `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR`. Please reference https://github.com/ARMmbed/mbedtls/issues/5596 comment in the commit message.

mprse commented:
This PR require the following PR in `psa-arch-tests` to be merged: https://github.com/bensze01/psa-arch-tests/pull/2.

cc @bensze01

superna9999 commented on library/psa_crypto.c:
```diff
@@ -4824,6 +4825,204 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+        goto cleanup;
+    }
+
+    while ( key_out_of_range )
```
```suggestion
    while( key_out_of_range )
```

mprse replied:
I fixed this spaces in my code, but I'm a bit confused looking at `psa_crypto.c` code. It seems that both forms are used.

superna9999 commented on library/psa_crypto.c:
```diff
@@ -4824,6 +4825,204 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+    while ( key_out_of_range )
+    {
+        /* 1. Draw a byte string of length ceiling(m/8) bytes. */
+        if ( ( status = psa_key_derivation_output_bytes( operation, *data, m_bytes ) ) != 0 )
```
```suggestion
        if( ( status = psa_key_derivation_output_bytes( operation, *data, m_bytes ) ) != 0 )
```

superna9999 commented on library/psa_crypto.c:
```diff
@@ -4824,6 +4825,204 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+            goto cleanup;
+
+        /* 2. If m is not a multiple of 8 */
+        if ( m % 8 != 0 )
```
```suggestion
        if( m % 8 != 0 )
```

superna9999 commented on library/psa_crypto.c:
```diff
@@ -4824,6 +4825,204 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+cleanup:
+    if( ret != 0 )
+        status = mbedtls_to_psa_error( ret );
+    if ( status != PSA_SUCCESS  ) {
```
```suggestion
    if( status != PSA_SUCCESS  ) {
```

superna9999 commented on library/psa_crypto.c:
```diff
@@ -4832,23 +5031,57 @@ static psa_status_t psa_generate_derived_key_internal(
+    if ( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
+    {
+        psa_ecc_family_t curve = PSA_KEY_TYPE_ECC_GET_FAMILY( slot->attr.type );
+        if ( PSA_ECC_FAMILY_IS_WEIERSTRASS( curve ) )
```
```suggestion
        if( PSA_ECC_FAMILY_IS_WEIERSTRASS( curve ) )
```

superna9999 commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6754,6 +6754,65 @@ exit:
+    size_t export_length;
+
+    ASSERT_ALLOC( export_buffer, export_buffer_size );
+    PSA_ASSERT( psa_crypto_init( ) );
```
```suggestion
    PSA_INIT( );
```

mprse replied:
I left this unchanged for constituency with other test cases.
This should be fixed separately as I think this PR has already become very big.

superna9999 commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6754,6 +6754,65 @@ exit:
+            &operation, base_key, alg,
+            input1->x, input1->len,
+            input2->x, input2->len,
+            PSA_KEY_DERIVATION_UNLIMITED_CAPACITY ) )
```
I'll suggest:
```C
    if( mbedtls_test_psa_setup_key_derivation_wrap(
            &operation, base_key, alg,
            input1->x, input1->len,
            input2->x, input2->len,
            PSA_KEY_DERIVATION_UNLIMITED_CAPACITY ) == 0 )
```

mprse replied:
Fixed.

superna9999 commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6754,6 +6754,65 @@ exit:
+exit:
+    mbedtls_free( export_buffer );
+    psa_key_derivation_abort( &operation );
+    psa_destroy_key( base_key );
```
```suggestion
    PSA_ASSERT( psa_destroy_key( base_key ) );
```

mprse replied:
I left this unchanged for constituency with other test cases. This is only test, but it would be good general change.
I think that this should be fixed separately as this PR has already become very big.

superna9999 commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6754,6 +6754,65 @@ exit:
+    mbedtls_free( export_buffer );
+    psa_key_derivation_abort( &operation );
+    psa_destroy_key( base_key );
+    psa_destroy_key( derived_key );
```
```suggestion
    PSA_ASSERT( psa_destroy_key( derived_key ) );
```

mpg commented on library/psa_crypto.c:
```diff
@@ -4999,7 +4999,7 @@ static psa_status_t psa_generate_derived_ecc_key_montgomery_helper(
             break;
         default:
-            /* should never happen */
+            return( PSA_ERROR_CORRUPTION_DETECTED );
```
Minor: the `break` on the next line seems redundant now.

mpg reviewed (changes requested):
Looks good to me, except that preprocessor conditions still do not seem to match [the last comment from Gilles](https://github.com/ARMmbed/mbedtls/pull/5139#pullrequestreview-899376626):
> Hence `defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR) || defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY)` is the right condition here.

Please resolve this (either by convincing Gilles or changing the code) and then I'll approve.

gilles-peskine-arm commented:
@mpg See https://github.com/ARMmbed/mbedtls/pull/5139#discussion_r818965361 . I'm convinced that the conditions are wrong, but it's a preexisting problem that's too hard to fix here and is now tracked as https://github.com/ARMmbed/mbedtls/issues/5596.

mpg reviewed (approved):
LGTM.

gilles-peskine-arm reviewed (approved):
Looks good to me (apart from [a continuation of a preexisting issue](https://github.com/ARMmbed/mbedtls/pull/5139#discussion_r818965361) which [I'm happy to defer](https://github.com/ARMmbed/mbedtls/issues/5596).

gilles-peskine-arm reviewed (changes requested):
Oops, no, we were forgetting: please add a changelog entry for the new feature!

mprse commented:
Added change log file.

mpg reviewed (approved):
LGTM

mprse commented:
CI passed.

mprse commented:
> CI passed.

Hmm I'm almost sure that on Friday I saw green CI. Now I see some strange failures (after adding change log file). Seems not related:

`No such property: pullRequest for class: groovy.lang.Binding`

`Scripts not permitted to use new java.util.NoSuchElementException java.lang.String`.

Restarted CI.

mpg commented:
Those errors are only on the "TF OpenCI", which is not stable yet. The existing CI ("Internal CI" and "Travis CI") is all green.
